test/e2e/grpc_test.go

```diff
@@ -92,13 +96,108 @@ func unaryTest(t *testing.T, resources *v1a1test.ResourceObjects, clients *test.
 	if err != nil {
 		t.Fatalf("Couldn't send request: %v", err)
 	}
-
-	if got.Msg != want.Msg {
+	if !strings.HasPrefix(got.Msg, want.Msg) {
 		t.Errorf("Response = %q, want = %q", got.Msg, want.Msg)
 	}
 }
 
-func streamTest(t *testing.T, resources *v1a1test.ResourceObjects, clients *test.Clients, host, domain string) {
+func autoscaleTest(t *testing.T, resources *v1a1test.ResourceObjects, clients *test.Clients, names test.ResourceNames, host, domain string) {
+	t.Helper()
+	t.Logf("Connecting to grpc-ping using host %q and authority %q", host, domain)
+
+	ctx := &testContext{
+		t:                 t,
+		clients:           clients,
+		resources:         resources,
+		names:             names,
+		targetUtilization: targetUtilization,
+	}
+	assertGRPCAutoscaleUpToNumPods(ctx, 1, 3, 60*time.Second, host, domain)
+	assertScaleDown(ctx)
+	assertGRPCAutoscaleUpToNumPods(ctx, 0, 3, 60*time.Second, host, domain)
+}
+
+func generateGRPCTraffic(t *testing.T, targetConcurrency int, host, domain string) (sync.Map, error) {
+	t.Helper()
+	var grp errgroup.Group
+	uniqueHost := sync.Map{}
+
+	for i := 0; i < targetConcurrency; i++ {
+		i := i
+		grp.Go(func() error {
+			conn, err := dial(host, domain)
+			if err != nil {
+				t.Fatalf("fail to dial: %v", err)
+			}
+			defer conn.Close()
+
+			pc := ping.NewPingServiceClient(conn)
+			want := &ping.Request{Msg: fmt.Sprintf("Hello! :%d", i)}
+
+			got, err := pc.Ping(context.Background(), want)
+			if err != nil {
+				return fmt.Errorf("Could not send request: %v", err)
+			}
+
+			if !strings.HasPrefix(got.Msg, want.Msg) {
+				return fmt.Errorf("Response = %q, want prefix = %q", got.Msg, want.Msg)
+			}
+
+			host := strings.TrimPrefix(got.Msg, want.Msg)
+			uniqueHost.Store(host, struct{}{})
+			return nil
+		})
+	}
+
+	if err := grp.Wait(); err != nil {
+		return uniqueHost, err
+	}
+	return uniqueHost, nil
+}
+
+func assertGRPCAutoscaleUpToNumPods(ctx *testContext, curPods, targetPods float64, duration time.Duration, host, domain string) {
+	ctx.t.Helper()
+	// Test succeeds when the number of pods meets targetPods.
+
+	// Relax the bounds to reduce the flakiness caused by sampling in the autoscaling algorithm.
+	// Also adjust the values by the target utilization values.
+
+	minPods := math.Floor(curPods/ctx.targetUtilization) - 1
+	maxPods := math.Ceil(targetPods/ctx.targetUtilization) + 1
+
+	stopChan := make(chan struct{})
+	var grp errgroup.Group
+
+	grp.Go(func() error {
+		uniqueHost, err := generateGRPCTraffic(ctx.t, 50, host, domain)
+		if err != nil {
+			return err
+		}
+		knownHosts := make(map[interface{}]interface{})
+
+		uniqueHost.Range(func(k, v interface{}) bool {
+			knownHosts[k] = v
+			return true
+		})
+		// This assertion checks whether responses are coming from multiple hosts
```

Comment:
rm commented code?

Reply:
refer comment : https://github.com/knative/serving/pull/6618#issuecomment-579909698

Reply:
Yeah, remove the code and then re-introduce it when the time comes.

Reply:
Well if we remove that code, we can remove all the uniqueHost map stuff with it. We only do that to facilitate this check.

Reply:
yes